[docs] Improved the main README of the SRT project (#2357)
Haivision/srt: README.md

  </p>
</details>

<details>
  <summary>Content Agnostic</summary>

  <p>

  Unlike some other streaming protocols that only support specific video and audio formats, SRT is payload agnostic. Because SRT operates at the network transport level, acting as a wrapper around your content, it can transport any type of video format, codec, resolution, or frame rate.

  </p>
</details>

<details>
  <summary>Easy Firewall Traversal with Rendezvous Mode</summary>

  <p>

  The handshaking process used by SRT supports outbound connections without the potential risks and dangers of permanent exterior ports being opened in a firewall, thereby maintaining corporate LAN security policies and minimizing the need for IT intervention.

  </p>
</details>

<details>
  <summary><a href="https://datatracker.ietf.org/doc/html/draft-sharabayko-srt-01#section-6">AES Encryption</a></summary>

  <p>

>   Using 128/192/256-bit AES encryption trusted by governments and organizations around the world, SRT ensures that valuable content is protected end-to-end from contribution to distribution so that no unauthorized parties can listen.

Reviewer: Feels like some supportive reference is missing... :)

  </p>
</details>

<details>
  <summary><a href="./docs/features/packet-filtering-and-fec.md">Forward Error Correction (FEC) and Packet Filter API</a></summary>

>   <p>

Reviewer 2: @ethouris It was written in HTML to take advantage of the collapsible <details> tag. Since the README is the "front face" of the repo, my opinion is that we should be less restrictive about the use of MarkDown in order to have the freedom to format the output to improve its appearance on the main page.

  For broadcasters, Stream ID is key to replacing RTMP for ingesting video streams, especially HEVC/H.265 content, into cloud service or CDNs that have a single IP socket (address + port) open for incoming video.

  </p>
</details>

## Getting Started with SRT

|                                                                                                                               |                                                                                    |                                                                                   |
|:-----------------------------------------------------------------------------------------------------------------------------:|:----------------------------------------------------------------------------------:|:---------------------------------------------------------------------------------:|

> | [The SRT API](./docs#srt-api-documents)                                                                                       | [Internet Draft](https://datatracker.ietf.org/doc/html/draft-sharabayko-srt-01)    | [Sample Apps](./docs#sample-applications)                                         |

Reviewer: ```suggestion
| [The SRT API](./docs#srt-api-documents)                                                                                       | [IETF Internet Draft](https://datatracker.ietf.org/doc/html/draft-sharabayko-srt-01)    | [Sample Apps](./docs#sample-applications)                                         |
```

- [Documentation on GitHub](./docs) with SRT API documents, features decsriptions, etc.
- The SRT Protocol Internet Draft: [Datatracker](https://datatracker.ietf.org/doc/draft-sharabayko-srt/) | [Latest Version](https://datatracker.ietf.org/doc/html/draft-sharabayko-srt-01) | [Latest Working Copy](https://haivision.github.io/srt-rfc/draft-sharabayko-srt.html) | [GitHub Repo](https://github.com/Haivision/srt-rfc)

## Build Instructions

[Linux (Ubuntu/CentOS)](./docs/build/build-linux.md) | [Windows](./docs/build/build-win.md) | [macOS](./docs/build/build-macOS.md) | [iOS](./docs/build/build-iOS.md) | [Android](./docs/build/build-android.md)

### Requirements

* C++03 or above compliant compiler.
* CMake 2.8.12 or above as a build system.
* OpenSSL 1.1 to enable encryption, otherwise build with [`-DENABLE_ENCRYPTION=OFF`](./docs/build/build-options.md#enable_encryption).
* Multithreading is provided by either of the following:
  * C++11: standard library (`std` by [`-DENABLE_STDCXX_SYNC=ON`](./docs/build/build-options.md#enable_stdcxx_sync) CMake option),

>   * C++03: Pthreads (for POSIX systems it's built in, for Windows there is a ported library).

Reviewer 3: Is building with pthread-win32 still supported?

Reviewer: Supported, but not encouraged.

Author: Yes, consulted with @maxsharabayko.

- [`lib-srt-utils`](https://github.com/mbakholdina/lib-srt-utils) - A Python library containing supporting code for running SRT tests based on an experiment configuration.

## Contributing

Anyone is welcome to contribute. If you decide to get involved, please take a moment to review the guidelines:

* [SRT Developer's Guide](docs/dev/developers-guide.md)
* [Contributing](CONTRIBUTING.md)
* [Reporting Issues](docs/dev/making-srt-better.md)

For information on contributing to the [Internet Draft](https://datatracker.ietf.org/doc/html/draft-sharabayko-srt-01) or to submit issues please go to the following [repo](https://github.com/Haivision/srt-rfc). The repo for contributing in [SRT CookBook](https://srtlab.github.io/srt-cookbook/) can be found [here](https://github.com/SRTLab/srt-cookbook/).

## License

By contributing code to the SRT project, you agree to license your contribution under the [MPLv2.0 License](LICENSE).

## Release History

- [Release notes](https://github.com/Haivision/srt/releases)
- [SRT versioning](./docs/dev/developers-guide.md#versioning)

> [^1]: The term “live streaming” refers to MPEG-TS style continuous data transmission with latency management. Live streaming based on segmentation and transmission of files like in HTTP Live Streaming (HLS) protocol as described in RFC8216 is not part of this use case. Note that SRT is content agnostic meaning that any type of data (not necessarily limited to MPEG-TS) can be transmitted via its payload.

Reviewer: ```suggestion
[^1]: The term “live streaming” refers (but not limited) to MPEG-TS style continuous data transmission with latency management. Live streaming based on segmentation and transmission of files like in HTTP Live Streaming (HLS) protocol as described in RFC8216 is not part of this use case. Note that SRT is content agnostic meaning that any type of data (not necessarily limited to MPEG-TS) can be transmitted via its payload.
```

Reviewer 3: I don't think it's required. It's said about "MPEG-TS style", not exclusively MPEG-TS format. Maybe this style could be explained differently, like "data sent with the same average rate as being produced and expected to be received". @stevomatthews maybe you have some better idea how to say that.

Reviewer 2: ```suggestion
[^1]: The term “live streaming” refers to continuous data transmission (MPEG-TS or equivalent) with latency management, which is to say where the average transmission rate is expected to be the same for the sender and the receiver. Live streaming based on segmentation and transmission of files like in HTTP Live Streaming (HLS) protocol as described in RFC8216 is not part of this use case. Note that SRT is content agnostic, meaning that any type of data (not necessarily limited to MPEG-TS) can be transmitted via its payload.
```

> [^1]: The term “live streaming” refers to continuous data transmission (MPEG-TS or equivalent) with latency management. Live streaming based on segmentation and transmission of files like in HTTP Live Streaming (HLS) protocol as described in RFC8216 is not part of this use case. File transmission in either message, or buffer mode should be considered in this case. See [Section 7. Best Practices and Configuration Tips for Data Transmission via SRT](https://datatracker.ietf.org/doc/html/draft-sharabayko-srt-01#section-7) of the Internet Draft for details. Note that SRT is content agnostic, meaning that any type of data can be transmitted via its payload.

Author: @stevomatthews The very last bit to review.

Reviewer 2: ```suggestion
[^1]: The term “live streaming” refers to continuous data transmission (MPEG-TS or equivalent) with latency management. Live streaming based on segmentation and transmission of files like in the HTTP Live Streaming (HLS) protocol (as described in RFC8216) is not part of this use case. File transmission in either message or buffer mode should be considered in this case. See [Section 7. Best Practices and Configuration Tips for Data Transmission via SRT](https://datatracker.ietf.org/doc/html/draft-sharabayko-srt-01#section-7) of the Internet Draft for details. Note that SRT is content agnostic, meaning that any type of data can be transmitted via its payload.
```

[![Build Status Linux and macOS][travis-badge]][travis]
[![Build Status Windows][appveyor-badge]][appveyor]

## What is SRT?

**Secure Reliable Transport (SRT)** is a transport protocol for ultra low (sub-second) latency live video and audio streaming¹, as well as for generic bulk data transfer. SRT is available as an open-source technology with the code on GitHub, published [Internet Draft](https://datatracker.ietf.org/doc/html/draft-sharabayko-srt-01), and a growing [community of SRT users](https://www.srtalliance.org/).

SRT is applied to contribution and distribution endpoints as part of a video stream workflow to deliver the best quality and lowest latency video at all times.

|               |                                                   |
| ------------- | ------------------------------------------------- |
| **S**ecure    | Encrypts video streams                            |
| **R**eliable  | Recovers from severe packet loss                  |
| **T**ransport | Dynamically adapts to changing network conditions |

> In live streaming configuration, the SRT protocol maintains a constant end-to-end latency. This allows the live stream’s signal characteristics to be recreated on the receiver side, reducing the need for buffering. As packets are streamed from source to destination, SRT detects and adapts to real-time network conditions between the two endpoints. It helps compensate for jitter and bandwidth fluctuations due to congestion over noisy networks.

Reviewer 2: ```suggestion
In live streaming configurations, the SRT protocol maintains a constant end-to-end latency. This allows the live stream’s signal characteristics to be recreated on the receiver side, reducing the need for buffering. As packets are streamed from source to destination, SRT detects and adapts to real-time network conditions between the two endpoints. It helps compensate for jitter and bandwidth fluctuations due to congestion over noisy networks.
```

SRT offers various error recovery mechanisms for minimizing the packet loss that is typical of Internet connections, of which Automatic Repeat reQuest (ARQ) is the primary method. With ARQ, when a receiver detects that a packet is missing it sends an alert to the sender requesting retransmission of this missing packet. [Forward Error Correction (FEC)](./docs/features/packet-filtering-and-fec.md) and Path Redundancy (or [Connection Bonding](./docs/features/bonding-intro.md) in SRT terminology) are also supported by the protocol.

SRT [implements AES encryption](https://datatracker.ietf.org/doc/html/draft-sharabayko-srt-01#section-6) to protect the payload of the media streams.

> <p align="right">Subscribe to the <a href="https://medium.com/innovation-labs-blog/tagged/secure-reliable-transport">Innovation Labs Blog on Medium </a> to learn more about the protocol</p>

Reviewer 2: ```suggestion
<p align="right">Subscribe to the <a href="https://medium.com/innovation-labs-blog/tagged/secure-reliable-transport">Innovation Labs Blog on Medium </a> to learn more about the protocol.</p>
```

    </td>
    <td style="width:33%">
      <p align="center">A comprehensive overview of the protocol with the deployment guidelines</p>
    </td>
    <td style="width:33%">
      <p align="center">Development notes on the SRT protocol</p>
    </td>
  </tr>
  <tr>
    <td style="width:33%">
      <p align="center"><a href="https://medium.com/innovation-labs-blog/tagged/secure-reliable-transport"><b>Innovation Labs Blog</b></a></p>
    </td>
    <td style="width:33%">
      <p align="center" valign="top"><a href="https://www.youtube.com/channel/UCr35JJ32jKKWIYymR1PTdpA"><b>SRTLab YouTube Channel</b></a></p>
    </td>
    <td style="width:33%">
      <p align="center"><a href="https://srtalliance.slack.com"><b>Slack</b></a></p>
    </td>
  </tr>
  <tr>
    <td style="width:33%">
      <p align="center">The blog on Medium with SRT related technical articles</p>
    </td>
    <td style="width:33%">
      <p align="center">Technical YouTube channel with useful videos</p>
    </td>
    <td style="width:33%">
      <p align="center">Slack channels to get the latest updates and ask questions</p>
    </td>
  </tr>

> </table>

Reviewer 2: I did a test and it actually looks better in MarkDown. I will convert and reapply here.

> </table>

Reviewer 2: ```
|                                                                                                                               |                                                                                    |                                                                                   |
|:-----------------------------------------------------------------------------------------------------------------------------:|:----------------------------------------------------------------------------------:|:---------------------------------------------------------------------------------:|
| [The SRT API](./docs#srt-api-documents)                                                                                       | [Internet Draft](https://datatracker.ietf.org/doc/html/draft-sharabayko-srt-01)    | [Sample Apps](./docs#sample-applications)                                         |
| Reference documentation for the SRT library API                                                                               | The SRT Protocol Internet Draft                                                    | Instructions for using test apps (`srt-live-transmit`, `srt-file-transmit`, etc.) |
| [SRT Technical Overview](https://github.com/Haivision/srt/files/2489142/SRT_Protocol_TechnicalOverview_DRAFT_2018-10-17.pdf)  | [SRT Deployment Guide](https://www.srtalliance.com/srt-deployment-guide/)          | [SRT CookBook](https://srtlab.github.io/srt-cookbook)                             |
| Early draft technical overview (precursor to the Internet RFC Draft)                                                          | A comprehensive overview of the protocol with deployment guidelines                | Development notes on the SRT protocol                                             |
| [Innovation Labs Blog](https://medium.com/innovation-labs-blog/tagged/secure-reliable-transport                               | [SRTLab YouTube Channel](https://www.youtube.com/channel/UCr35JJ32jKKWIYymR1PTdpA) | [Slack](https://srtalliance.slack.com)                                            |
| The blog on Medium with SRT-related technical articles                                                                        | Technical YouTube channel with useful videos                                       | Slack channels to get the latest updates and ask questions                        |

```

> ### Further Documentation

Reviewer 2: ```suggestion
### Additional Documentation
```

<p align="right"><em>To learn more about the protocol subscribe to the <a href="https://medium.com/innovation-labs-blog/tagged/secure-reliable-transport">Innovation Labs Blog</a> on &nbsp;<img alt="slack logo" src="https://upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Medium_%28website%29_logo.svg/500px-Medium_%28website%29_logo.svg.png" width="80"></em></p>

<p align="right"><em>To ask a question <a href="https://slackin-srtalliance.azurewebsites.net">join the conversation</a> in the <b>#development</b> channel on &nbsp;<a href="https://srtalliance.slack.com"><img alt="slack logo" src="https://github.com/stevomatthews/srt/blob/master/docs/images/Slack_RGB2.svg" width="60"></a></em></p>

## Features

> :point_down: Click on the &#9658; button to expand a feature description.

<details>
  <summary>Pristine Quality and Reliability</summary>

  <p>

  No matter how unreliable your network, SRT can recover from severe packet loss and jitter, ensuring the integrity and quality of your video streams.

  </p>
</details>

<details>
  <summary>Low Latency</summary>

  <p>

>   SRT’s stream error correction is configurable to accommodate a user’s deployment conditions. Leveraging real-time IP communications development to extend traditional network error recovery practices, SRT delivers media with significantly lower latency than TCP/IP, while offering the speed of unreliable UDP transmission without the disadvantage.

Reviewer 2: latency than TCP/IP, while offering the speed of UDP transmission with greatly improved reliability.

  The first plugin created as an example of what can be achieved with the packet filter API is for Forward Error Correction (FEC) which, in certain use cases, albeit very rarely, where there are very long-distance links and an excellent quality connection, can offer slightly lower latency than Automatic Repeat reQuest (ARQ). This plugin allows users to compare ARQ and FEC and allow three different modes to be run:

   - ARQ only – retransmit of lost packets,
   - FEC only – provides the overhead in the bandwidth needed for recovery on the receiver side,

>    - FEC and ARQ – if FEC doesn’t manage to recover a packet, you can still retransmit.

Reviewer 2: - ARQ only
   - FEC only
   - FEC and ARQ

For more information, see [SRT Packet Filtering & FEC](/docs/features/packet-filtering-and-fec.md)

Reviewer 2: Okay then. Some minor edits:

   - ARQ only – retransmits lost packets,
   - FEC only – provides the overhead needed for FEC recovery on the receiver side,
   - FEC and ARQ – retransmits lost packets that FEC fails to recover a packet.

Author: @stevomatthews FEC and ARQ – retransmits lost packets that FEC fails to recover ?

Reviewer 2: Oops . . . yes, I missed that. Thanks!

  </p>
</details>

<details>
  <summary><a href="./docs/features/bonding-intro.md">Connection Bonding</a></summary>

  <p>

>   Similar to SMPTE-2022-7 over managed networks, Connection Bonding adds seamless stream protection and hitless failover to the SRT protocol. This type of hitless protection and failover technology relies on more than one IP network path to prevent disruption to live video streams, in the event of network congestion or outages, by maintaining continuity of service.

Reviewer 2: Similar to SMPTE-2022-7 over managed networks, Connection Bonding adds seamless stream protection and hitless failover to the SRT protocol. This technology relies on more than one IP network path to prevent disruption to live video streams in the event of network congestion or outages, maintaining continuity of service.

>   This is done by introducing the [socket groups](./docs/features/socket-groups.md) in [SRT v1.5](https://github.com/Haivision/srt/releases/tag/v1.5.0). The general concept of socket groups means having a group that contains multiple sockets and that one operation for sending one data signal is applied to the group, compared to single sockets inside the group which will take over this operation and do what is necessary to deliver the signal to the receiver.

Reviewer 2: This is accomplished using the [socket groups](./docs/features/socket-groups.md) introduced in [SRT v1.5](https://github.com/Haivision/srt/releases/tag/v1.5.0). The general concept of socket groups means having a group that contains multiple sockets, where one operation for sending one data signal is applied to the group. Single sockets inside the group will take over this operation and do what is necessary to deliver the signal to the receiver.

  * C++03: Pthreads (for POSIX systems it's built in, for Windows there is a ported library).
* Tcl 8.5 (optional, used by `./configure` script or use CMake directly).

### Build Options

> For detailed descriptions of the build system and options, please read [SRT Build Options](./docs/build-options.md) document.

Reviewer 2: ```suggestion
For detailed descriptions of the build system and options, please read the [SRT Build Options](./docs/build-options.md) document.
```

## Sample Applications and Tools

> The current repo provides [sample applications](./apps) and [code examples](./examples) that demonstate the usage of SRT library API and should not be considered as "ready for production". Among them are `srt-live-transmit`, `srt-file-transmit`, and other applications. The respective documentation can be found [here](./docs#sample-applications).

Reviewer 2: ```suggestion
The current repo provides [sample applications](./apps) and [code examples](./examples) that demonstrate the usage of the SRT library API. Among them are `srt-live-transmit`, `srt-file-transmit`, and other applications. The respective documentation can be found [here](./docs#sample-applications). Note that all samples are provided for instructional purposes, and should not be used in a production environment.
```

> There is also an [`srt-xtransmit`](https://github.com/maxsharabayko/srt-xtransmit) test utility that is actively used for internal testing and performance evaluation. It supports dummy payload generation, traffic routing and other useful things. Please go and check the respective repo for details.

Reviewer 2: ```suggestion
The [`srt-xtransmit`](https://github.com/maxsharabayko/srt-xtransmit) utility is actively used for internal testing and performance evaluation. Among other features it supports dummy payload generation and traffic routings. Additional details are available in the  [`srt-xtransmit`](https://github.com/maxsharabayko/srt-xtransmit) repo itself.
```

- [`lib-srt-utils`](https://github.com/mbakholdina/lib-srt-utils) - A Python library containing supporting code for running SRT tests based on an experiment configuration.

## Contributing

Anyone is welcome to contribute. If you decide to get involved, please take a moment to review the guidelines:

* [SRT Developer's Guide](docs/dev/developers-guide.md)
* [Contributing](CONTRIBUTING.md)
* [Reporting Issues](docs/dev/making-srt-better.md)

To contribute in the [Internet Draft](https://datatracker.ietf.org/doc/html/draft-sharabayko-srt-01) or submit issues, please go to the following [repo](https://github.com/Haivision/srt-rfc).

If you have any questions, please feel free to ask in the <b>#development</b> channel on <a href="https://srtalliance.slack.com">Slack</a>.</p>

## License

By contributing code to the SRT project, you agree to license your contribution under the [MPLv2.0 License](LICENSE).

## Release History

- [Release notes](https://github.com/Haivision/srt/releases)

> - The info about [SRT versioning](./docs/dev/developers-guide.md#versioning)

Reviewer 2: ```suggestion
- [SRT versioning](./docs/dev/developers-guide.md#versioning)
```
